File: 526_Beautiful_Arrangement.py
class Solution:
    def countArrangement(self, N: int) -> int:
        def helper(nums, start):
            if start == 0:
                return 1
            count = 0
            for i in range(start, 0, -1):
                nums[start], nums[i] = nums[i], nums[start]
                if nums[start] % start == 0 or start % nums[start] == 0:
                    count += helper(nums, start - 1)
                nums[start], nums[i] = nums[i], nums[start]
            return count
        return helper([n for n in range(N + 1)], N)

File: 526_Beautiful_Arrangement.cpp
class Solution {
private:
    int helper(vector<int>& p, int start) {
        if (start == 0) return 1;
        int cnt = 0;
        for (int i = start; i > 0; --i) {
            swap(p[i], p[start]);
            if (start % p[start] == 0 || p[start] % start == 0)
                cnt += helper(p, start - 1);
            swap(p[i], p[start]);
        }
        return cnt;
    }
public:
    int countArrangement(int n) {
        vector<int> p (n + 1);
        for (int i = 0; i < n + 1; ++i) p[i] = i;
        return helper(p, n);
    }
};
